plugin/path_checkouts.go

```diff
@@ -5,11 +5,349 @@ import (
+			logical.ReadOperation: &framework.PathOperation{
+				Callback: b.operationSetCheckIn,
+				Summary:  "Check service accounts in to the library.",
+			},
+		},
+		HelpSynopsis: `Check service accounts in to the library.`,
+	}
+}
+
+func (b *backend) operationSetCheckIn(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+	return b.handleCheckIn(ctx, req, fieldData, false)
+}
+
+func (b *backend) pathSetManageCheckIn() *framework.Path {
+	return &framework.Path{
+		Pattern: libraryPrefix + "manage/" + framework.GenericNameRegex("name") + "/check-in$",
+		Fields: map[string]*framework.FieldSchema{
+			"name": {
+				Type:        framework.TypeLowerCaseString,
+				Description: "Name of the set.",
+				Required:    true,
+			},
+			"service_account_names": {
+				Type:        framework.TypeCommaStringSlice,
+				Description: "The username/logon name for the service accounts to check in.",
+			},
+		},
+		Operations: map[logical.Operation]framework.OperationHandler{
+			logical.ReadOperation: &framework.PathOperation{
+				Callback: b.operationSetManageCheckIn,
+				Summary:  "Check service accounts in to the library.",
+			},
+		},
+		HelpSynopsis: `Force checking service accounts in to the library.`,
+	}
+}
+
+func (b *backend) operationSetManageCheckIn(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+	return b.handleCheckIn(ctx, req, fieldData, true)
+}
+
+func (b *backend) handleCheckIn(ctx context.Context, req *logical.Request, fieldData *framework.FieldData, overrideCheckInEnforcement bool) (*logical.Response, error) {
+	setName := fieldData.Get("name").(string)
+
+	serviceAccountNamesRaw, serviceAccountNamesSent := fieldData.GetOk("service_account_names")
+	var serviceAccountNames []string
+	if serviceAccountNamesSent {
+		serviceAccountNames = serviceAccountNamesRaw.([]string)
+	}
+
+	set, err := readSet(ctx, req.Storage, setName)
+	if err != nil {
+		return nil, err
+	}
+	if set == nil {
+		return logical.ErrorResponse(`"%s" doesn't exist`, setName), nil
+	}
+
+	// If check-in enforcement is overridden or disabled at the set level, we should consider it disabled.
+	disableCheckInEnforcement := overrideCheckInEnforcement || set.DisableCheckInEnforcement
+
+	// Track the service accounts we check in so we can include it in our response.
+	var checkIns []string
+
+	// Build and validate a list of service account names that we will be checking in.
+	var cantCheckInErrs error
+	if len(serviceAccountNames) == 0 {
+		// It's okay if the caller doesn't tell us which service accounts they
+		// want to check in as long as they only have one checked out.
+		// We'll assume that's the one they want to check in.
+		toCheckIn := make(map[string]*locksutil.LockEntry)
+		for _, setServiceAccount := range set.ServiceAccountNames {
+			// This is called in a func to avoid calling defer in a loop.
+			if err := func() error {
```

Comment:
I do agree the code is less than beautiful there. I did also play with moving it out because I feel the same way.

When I did that, it was still weird. The reason is, there are items from outside the loop that are being used inside it (`toCheckIn` and `releaseLock`). If I move it to its own function, I need to either pass those things in repeatedly, or I need to return the stuff that will effect them. That ends up looking like a strange, poorly-encapsulated function. 

Since neither was clearly better, I decided to keep the function inline because there's been a general preference for inline code on the Vault team.

Reply:
This is no longer in a function because locks are now at the set level, so that has the by-product of solving this.